pingcap/tidb-operator
# add MaxSkew, MinDomains and NodeAffinityPolicy for TopologySpreadConstraint

### What problem does this PR solve?


close #5618

### What is changed and how does it work?


add some fields of `corev1.TopologySpreadConstraint`, but we can't inline this struct directly now, as it will change some __required__ behavior of the CRD.

![image](https://github.com/pingcap/tidb-operator/assets/3183039/43f9cc5e-f5b3-4394-8628-97584d4f6a3c)


### Code changes


### Tests


  1. add new fields in `topologySpreadConstraints`
    ```yaml
    topologySpreadConstraints:
    - topologyKey: kubernetes.io/hostname
      maxSkew: 2
      minDomains: 3
      nodeAffinityPolicy: Ignore
    ```
  2. check Pod has these fields after upgraded
    ```yaml
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            app.kubernetes.io/component: tikv
            app.kubernetes.io/instance: basic
            app.kubernetes.io/managed-by: tidb-operator
            app.kubernetes.io/name: tidb-cluster
        maxSkew: 2
        minDomains: 3
        nodeAffinityPolicy: Ignore
        topologyKey: kubernetes.io/hostname
        whenUnsatisfiable: DoNotSchedule
    ```

### Side effects


### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
add MaxSkew, MinDomains and NodeAffinityPolicy for TopologySpreadConstraint
```
